Full Relay Read/Write Support

buttercat1791 opened the pull request:
Fully support reading from and writing to relays.

## Description

- Create a `Filters` struct for creating relay query messages.
- Define an `ISigner` interface for handling event signing.
- Add support to `NostrService` for reading from relays.
    - `NostrService::queryRelays` sends a `Filters` object to the relay, parses response messages as the relay sends them, and passes the parsed `Event` objects to a caller-provided handler function.
    - `NostrService` generates a unique subscription ID for each filter, and maintains a list of active subscriptions.
    - The `NostrService::closeSubscription` and `NostrService::closeSubscriptions` methods allow the caller to close open relay subscriptions.

## Testing

Unit tests are provided for the newly-added methods in `NostrService`.

Currently, `ISigner` is mocked for tests.

## Notes for Reviewer

The `ISigner` interface is defined, since it is needed to sign events to send to relays, but a full implementation is not provided in this PR.  A signer will be added as part of a future PR.

buttercat1791 committed:
Match event.cpp formatting to rest of project

buttercat1791 committed:
Add validation on Event serialization

buttercat1791 committed:
Pass references into lambda

Unit tests for event publishing currently fail due to event validation.  A signer will need to be implemented before tests pass.

buttercat1791 committed:
Add a filters struct for relay queries

buttercat1791 committed:
Generate a valid ID while serializing an event

buttercat1791 committed:
Tweak variable names in Filters::serialize()

buttercat1791 committed:
Use SHA256 hashing function for OpenSSL >= 3.0

buttercat1791 committed:
Include SHA header for events

buttercat1791 committed:
Send events as arrays

buttercat1791 committed:
Serialize relay query filters into a JSON array

buttercat1791 committed:
Open relay subscriptions for a filter request

buttercat1791 committed:
Add service methods to close filter subscriptions

buttercat1791 committed:
Provide methods to close all open subscriptions

buttercat1791 committed:
Define a receive method on the WebSocket interface

buttercat1791 committed:
Begin defining relay message handling

buttercat1791 committed:
Provide option to store events for async retrieval

buttercat1791 committed:
Use smart pointers

buttercat1791 committed:
Namespace pointer declaration

buttercat1791 committed:
Implement receive method in WebSocket client

buttercat1791 committed:
Declare a signer interface

buttercat1791 committed:
Move relay payload parsing into NostrService

Preserve separation of concerns.

buttercat1791 committed:
Get smarter with pointers so tests pass

buttercat1791 committed:
Replace Event::deserialize with static methods

buttercat1791 committed:
Use shared pointers for filters and events

buttercat1791 committed:
Add a unit test for NostrService::QueryRelays

buttercat1791 committed:
Test provided handlers in queryRelays

buttercat1791 committed:
Remove default event handling

Caching events and fetching them in batches is out of scope for NostrService.  In the future, an additional service should be added to the library that handles local event caching and provides some default handlers for incoming messages from relays.

buttercat1791 committed:
Handle relay response messages

These changes do not yet have unit tests.

buttercat1791 committed:
Update existing unit tests for recent code changes

All preexisting unit tests now pass and test for the correct behavior.

Silberengel commented on src/event.cpp:
```diff
@@ -1,38 +1,111 @@
+{
+    Event event;
+
+    event.id = j["id"];
```
Unchecked access with [] operator is often a source of failure. Better to do check access with default or add some try catch. ( https://json.nlohmann.me/features/element_access/ )

Silberengel commented on include/nostr.hpp:
```diff
@@ -76,16 +157,80 @@ class NostrService
-    */
-    std::tuple<RelayList, RelayList> publishEvent(Event event);
+     */
+    std::tuple<RelayList, RelayList> publishEvent(std::shared_ptr<Event> event);
```
Good! :-)

Silberengel commented on include/nostr.hpp:
```diff
@@ -76,16 +157,80 @@ class NostrService
+     * to which relays the message was sent successfully, and which relays failed to receive the
+     * message.
+     */
+    std::tuple<RelayList, RelayList> closeSubscriptions(RelayList relays);
```
Not sure if this RelayList could/should be const.

Silberengel commented on src/nostr_service.cpp:
```diff
@@ -134,13 +197,209 @@ tuple<RelayList, RelayList> NostrService::publishEvent(Event event)
+        promise<tuple<string, bool>> eosePromise;
+        requestFutures.push_back(move(eosePromise.get_future()));
+
+        this->_client->send(
```
If this send fails with an exception, let's say on the first relay, what happens with the subsequent relays?

buttercat1791 committed:
Improve error handling around JSON parsing

buttercat1791 committed:
Refine error handling on wss send

buttercat1791 committed:
Take send out of try-catch

The client::IWebSocketClient::send method should catch errors and return false if anything goes wrong.

buttercat1791 committed:
Unit test rejected events

buttercat1791 committed:
Add unit tests for batch queries

Also add an equality operator for nostr::Event

finrodfelagund97 committed:
configure vcpkg build

finrodfelagund97 committed:
fix include sha.h

finrodfelagund97 committed:
use uuid_v4 to generate faster UUIDs than Boost

finrodfelagund97 committed:
put binary outputs at the level of GitRepublic

finrodfelagund97 committed:
use namespaces instead of using specific variables

buttercat1791 committed:
Begin switching to FetchContent for deps

buttercat1791 committed:
Configure for Linux builds

- Use vcpkg for most dependency management.
- Manually include uuid_v4.
- Update README with prerequisites and build instructions.
- Support subproject and standalone builds.

buttercat1791 committed:
README instructions for testing

buttercat1791 committed:
Rename project to 'aedile'

buttercat1791 committed:
Ensure first queryRelay unit test passes

buttercat1791 committed:
Add build and test presets for CMake

buttercat1791 committed:
Give example commands for build and test

buttercat1791 committed:
Add unit test for queryRelays with callbacks

buttercat1791 committed:
Close relays after batch query and update unit test

buttercat1791 committed:
Test closing subscriptions

buttercat1791 committed:
Add details to README

buttercat1791 committed:
Update and test methods for closing subscriptions

buttercat1791 committed:
Remove 'RelayList' type alias
